helm/nginx-ingress-controller-app/values.yaml

```diff
@@ -26,13 +25,15 @@ configmap:
   # Increase hash table size to allow more server names for stability reasons
   server-name-hash-bucket-size: "1024"
   server-tokens: "false"
-  worker-processes: "4"
+  worker-processes: "1"
   use-forwarded-headers: "true"
 
 controller:
   name: nginx-ingress-controller
   k8sAppLabel: nginx-ingress-controller
 
+  replicaCount: 2
```

Comment:
@pipo02mix @puja108 @rossf7 recently GS AWS worker instance type with 2CPUs was removed and now min instance type has 4CPUs, still ASG allowed min and max remained to be 1. With such minimal test/playground/learning or whatever the use case for max single worker node clusters is (trying to discover that via https://gigantic.slack.com/archives/C0251MXL5/p1582671214019900), 2 nginx replicas with 2CPU each couldn't fit on such minimal single node 4CPU tenant cluster, especially not together with all the other pre-installed managed workloads.

Maybe we can consider increasing min worker count to 2.

Still, assuming we won't increase minimum worker count to 2, I'm thinking to either
- reduce the nginx replica count default to 1, and HPA minReplicas to 1 too, hopefully at least 1 can get scheduled on minimal single worker node cluster; con is that even on production environments nginx Deployment could be dropped by HPA to single replica only, so risky setup with no redundancy; or alternatively to
- keep 2 as the default replica count, tolerating that not both of them would be scheduled, I'd test that at least one does get scheduled; both of them being scheduled and nothing else of pre-installed components I think is not currently possible since nginx gets installed relatively late in cluster init; customer could be instructed to manually customize/optimize resource requests and autoscale settings in this non-productive corner case; pro would be that more important and more common non-idle usage case would be better covered with automation out of the box.

Longer term we could figure out supported use cases / usage profiles, if we can have them categorized to few T-shirt sizes determined based on resources assigned to the cluster (number range of workers and worker size / instance type), cluster-operator could have that resources -> profile mapping logic and share outcome in cluster values ConfigMap the determined cluster usage profile/category, based on profile apps can adjust default resource requests and redundancy settings, which would hopefully minimize need to have resource/scaling customized for the supported usage profiles.

What do you think?

Reply:
My understanding (which is limited) is that you can create single node clusters but they are not fully supported. 

The 1st option of reducing to 1 replica seems simpler. There is a risk of downtime to the ingress but if you've configured a single node cluster you've increased your risk of downtime.

Reply:
can we do an intermediate solution like setting a min value depending on the size of the cluster?

-> cluster single NP with 1 min 3 max workers -> HPA IC min 1 
-> cluster single NP with 3 min 10 max workers -> HPA IC min 2
-> cluster single NP with 10 min 50 max workers -> HPA IC min 4

Problem is the cluster scaling actions should modify the values on HPA of some components (here IC) but IMO it would be the ideal scenario. I talked to a customer with big node pool cluster and they were surprised the HPA min value for CoreDNS was 2 replicas. I think condition the min (or max) to cluster size is the ideal. Can we do it somehow?

Reply:
@rossf7 thanks for your feedback, on single node cluster use-case and support

It makes no sense to have min HPA higher than default replica count, as HPA would scale up immediately to its min.
I think defaulting to 1 replica (so also 1 min HPA) is availability risk even for tenant clusters which have more than 1 worker node - nginx, or node will crash or get recycled; we saw with recent incidents with Pods stuck Terminating for ages, it takes time for replacement to come up, so 1 replica is risk even in multi-node clusters.

Therefore I'm more for defaulting to 2 replicas, letting 1 hang unscheduled in playground single small node clusters.
Will go with that for now (so keep settings as is) and test in a WIP release even on a single node cluster with smallest instance type - can always come back and make adjustments in a follow up patch release.

Reply:
@pipo02mix yes, mentioned earlier the idea with having cluster-operator do t-shirt sizes / usage profile estimation, estimation would be based on what customer assigned to the tenant cluster (scaling range i.e. instance count and instance type / capacity), estimated profile would be saved/updated in (tenant) cluster values ConfigMap that would be provided to every installed app, and app charts could have mapping of t-shirt sizes to app configuration (replica count, resource assignment, HPA range, ...)

Start simple with 2-3 levels/profiles - think anything is better than current single one-size fits all, apps are not smart enough to deal with the range of clusters/usage, we practically can't apply both HPA and VPA at the same time to a single app.

Reply:
Tested HPA enabled nginx with 2 initial replicas, and HPA with 2min replicas on smallest single node cluster we currently support on aws, using 11.0.1 platform release, and installing nginx App as optional/ondemand without any adjustments - other things started as expected, while there was single running nginx replica and 2 Pending - HPA was trying to scale out since the desired replica count was lower than the minimum replica count.

Problem with this configuration is that we'll get paged about nginx Deployment not being satisfied. For this reason (to cover smallest cluster without paging) I'll reduce initial replica count to 1 and also HPA min to 1.